File: corporate-reporting/src/types/CustomReportingEntity.ts
import { ICustomReportingEntityType } from "./CustomReportingEntityType";
import { ICustomReportingEntityUpdate } from "./CustomReportingEntityUpdate";
import { IDirectorate } from "./Directorate";
import { IEntity } from "./Entity";
import { IPartnerOrganisation } from "./PartnerOrganisation";
import { IProject } from "./Project";
import { IReportingEntityWithDeliveryDates, ReportingEntityWithDeliveryDates } from "./ReportingEntity";
import { IUser } from "./User";

export interface ICustomReportingEntity extends IReportingEntityWithDeliveryDates {
    ReportingEntityTypeID: number;
    LeadUserID: number;
    DirectorateID: number;
    ProjectID: number;
    PartnerOrganisationID: number;
    TargetPerformanceUpperLimit: number | string;
    TargetPerformanceLowerLimit: number | string;
    MeasurementUnitID: number;

    MeasurementUnit?: IEntity;
    ReportingEntityType?: ICustomReportingEntityType;
    Directorate?: IDirectorate;
    Project?: IProject;
    PartnerOrganisation?: IPartnerOrganisation;
    LeadUser?: IUser;
    ReportingEntityUpdates?: ICustomReportingEntityUpdate[];
}

export class CustomReportingEntity extends ReportingEntityWithDeliveryDates implements ICustomReportingEntity {
    public ReportingEntityTypeID = null;
    public LeadUserID = null;
    public DirectorateID = null;
    public ProjectID = null;
    public PartnerOrganisationID = null;
    public TargetPerformanceUpperLimit = '';
    public TargetPerformanceLowerLimit = '';
    public MeasurementUnitID = null;

    constructor(reportingEntityTypeId: number) {
        super();
        this.ReportingEntityTypeID = reportingEntityTypeId;
    }
}

File: ORB.Core/Models/CustomReportingEntity.cs
using System;
using System.Collections.Generic;

namespace ORB.Core.Models
{
    public class CustomReportingEntity : ReportingSubEntity
    {
        public CustomReportingEntity()
        {
            ReportingEntityUpdates = new HashSet<CustomReportingEntityUpdate>();
        }

        public int ReportingEntityTypeID { get; set; }
        public int? DirectorateID { get; set; }
        public int? ProjectID { get; set; }
        public int? PartnerOrganisationID { get; set; }
        public DateTime? CreatedDate { get; set; }
        public int? MeasurementUnitID { get; set; }
        public decimal? TargetPerformanceUpperLimit { get; set; }
        public decimal? TargetPerformanceLowerLimit { get; set; }
        public DateTime? BaselineDate { get; set; }
        public DateTime? ForecastDate { get; set; }
        public DateTime? ActualDate { get; set; }
        public IDictionary<string, object> Properties { get; set; }

        public virtual CustomReportingEntityType ReportingEntityType { get; set; }
        public virtual Directorate Directorate { get; set; }
        public virtual Project Project { get; set; }
        public virtual PartnerOrganisation PartnerOrganisation { get; set; }
        public virtual MeasurementUnit MeasurementUnit { get; set; }
        public virtual ICollection<CustomReportingEntityUpdate> ReportingEntityUpdates { get; set; }
    }
}
